anomdetect/src/main/java/com/expedia/adaptivealerting/anomdetect/DetectorMapper.java

```diff
@@ -15,43 +15,95 @@
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
 
 /**
  * Entry into the Adaptive Alerting runtime. Its job is find for any incoming {@link MetricData} the corresponding set
  * of mapped detectors, creating a {@link MappedMetricData} for each.
  */
-@RequiredArgsConstructor
 @Slf4j
 public class DetectorMapper {
-
+    private AtomicLong indexSize;
+    private DetectorMapperCache cache;
+    private AtomicLong lastElasticLookUpLatency = new AtomicLong(-1);
     @Getter
     @NonNull
     private DetectorSource detectorSource;
 
-    /**
-     * Maps an {@link MetricData} to its corresponding set of {@link MappedMetricData}s.
-     *
-     * @param metricData MetricData to map.
-     * @return The corresponding set of {@link MappedMetricData}s: one per detector.
-     */
-    public Set<MappedMetricData> map(MetricData metricData) {
-        notNull(metricData, "metricData can't be null");
-        return detectorSource
-                .findDetectorUuids(metricData.getMetricDefinition())
-                .stream()
-                .map(detectorUuid -> new MappedMetricData(metricData, detectorUuid))
-                .collect(Collectors.toSet());
+
+    public DetectorMapper(DetectorSource detectorSource) {
+        assert detectorSource!=null;
+        this.detectorSource = detectorSource;
+        this.indexSize = Metrics.gauge("index.size", new AtomicLong(0));
+        this.cache = new DetectorMapperCache();
+    }
+
+    public List<Detector> getDetectorsFromCache(MetricData metricData) {
+        String cacheKey = CacheUtil.getKey(metricData.getMetricDefinition().getTags().getKv());
+        return cache.get(cacheKey);
     }
+
+    public boolean isSuccessfulDetectorMappingLookup(List<Map<String, String>> cacheMissedMetricTags) {
+
+        log.info("Mapping-Cache: lookup for {} metrics", cacheMissedMetricTags.size());
+        DetectorMatchResponse matchingDetectorMappings = detectorSource.findMatchingDetectorMappings(cacheMissedMetricTags);
+
+        if (matchingDetectorMappings != null) {
+
+            lastElasticLookUpLatency.set(matchingDetectorMappings.getLookupTimeInMillis());
+            Map<Integer, List<Detector>> groupedDetectorsByIndex = matchingDetectorMappings.getGroupedDetectorsBySearchIndex();
+
+            //populate cache and result map
+            groupedDetectorsByIndex.forEach((index, detectors) -> {
+                String cacheKey = CacheUtil.getKey(cacheMissedMetricTags.get(index));
+                if (!detectors.isEmpty()) {
+                    cache.put(cacheKey, detectors);
+                }
+            });
+
+            Set<Integer> searchIndexes = groupedDetectorsByIndex.keySet();
+            indexSize.set(searchIndexes.size());
+
+//For metrics with no matching detectors, set matching detectors to empty in cache to avoid repeated cache miss
+            final AtomicInteger i = new AtomicInteger(0);
```

Comment:
Not seeing why this is an `AtomicInteger` as opposed to a normal `int`. Our access to this variable appears to be single-threaded.

Reply:
I think this was added because forEach  requires finals,  doesn't take just int. I am replacing this wit for-each loop